presidential-elections/utils/model.py

```diff
@@ -449,50 +466,299 @@ def build_model(
+            # only place where poll results enter the model.
+            # --------------------------------------------------------
+
+            latent_mu = (
+                party_intercept
+                + election_party_intercept[data_containers["election_idx"]]
+                + party_time_effect[data_containers["countdown_idx"]]
+                + election_party_time_effect[
+                    data_containers["countdown_idx"], :, data_containers["election_idx"]
+                ]
+                + aet.dot(
+                    data_containers["stdz_unemp"][:, None], unemployment_effect[None, :]
+                )
+            )
+            latent_mu = latent_mu + non_competing_parties["polls_additive"]
+            pm.Deterministic(
+                "latent_popularity",
+                aet.nnet.softmax(latent_mu),
+                dims=("observations", "parties_complete"),
+            )
+            noisy_mu = (
+                latent_mu
+                + poll_bias[None, :]  # let bias vary during election period?
+                + house_effects[data_containers["pollster_idx"]]
+                + house_election_effects[
+                    data_containers["pollster_idx"], :, data_containers["election_idx"]
+                ]
+                * non_competing_parties["polls_multiplicative"]
+            )
+
+            noisy_popularity = pm.Deterministic(
+                "noisy_popularity",
+                aet.nnet.softmax(noisy_mu),
+                dims=("observations", "parties_complete"),
+            )
+
+            # The concentration parameter of a Dirichlet-Multinomial distribution
+            # can be interpreted as the effective number of trials.
+            #
+            # The mean (1000) is thus taken to be roughly the sample size of
+            # polls, and the standard deviation accounts for the variation in
+            # sample size.
+            concentration_polls = pm.InverseGamma(
+                "concentration_polls", mu=1000, sigma=200
             )
 
-            concentration = pm.InverseGamma("concentration", mu=1000, sigma=200)
             pm.DirichletMultinomial(
                 "N_approve",
-                a=concentration * noisy_popularity,
+                a=concentration_polls * noisy_popularity,
                 n=data_containers["observed_N"],
                 observed=data_containers["observed_polls"],
                 dims=("observations", "parties_complete"),
             )
+
+            # --------------------------------------------------------
+            #                    ELECTION RESULTS
+            #
+            # In this section we use the variables defined before to model the
+            # political families' latent popularity and how it translates into
+            # results the day of the election.
+            #
+            # Fundamental data only enter the model here; poll results enter
```

Comment:
> Fundamental data only enter the model here;

Mmmh no, fundamentals are already used in `latent_mu` above. This is exactly the same regression, except only at time 0. It's just that that timestep is informed by election results, whereas previous timesteps are informed by polls

Reply:
True, I meant election results. Good catch.